app/models/remote_identity.rb

```diff
@@ -30,8 +30,9 @@
 
 class RemoteIdentity < ApplicationRecord
   belongs_to :user
-  belongs_to :oauth_client, class_name: "OAuthClient"
+  belongs_to :auth_source, polymorphic: true
+  belongs_to :integration, polymorphic: true
 
-  validates :user, uniqueness: { scope: :oauth_client }
-  validates :origin_user_id, :user, :oauth_client, presence: true
+  validates :user, uniqueness: { scope: %i[auth_source integration] }
```

Comment:
:yellow_circle:  I am really not sure what's the right thing to do here.

I am pretty sure that there will be code (possibly touched in this PR), that will _assume_ there is exactly one user per integration. Because uniqueness per storage is what we effectively had before.

Thus it would make sense to keep it that way:

```suggestion
  validates :user, uniqueness: { scope: :integration }
```

However, this would be a problem for users migrating to SSO if either of these is true:
* code relies on looking up the user with their current auth source
* the origin user id is different between SSO and the old authentication

On the other hand: Allowing to migrate has been extracted to a separate feature, so being more restrictive might be okay for now.

What do you think?

Reply:
I think we should go with `auth_source integration` scope. I know one place to update where we should be explicit about auth_source.  That is `modules/storages/app/common/storages/peripherals/storage_interaction/nextcloud/util.rb`. I suggest there we have something like:
``` ruby
              when :oauth_user_token
                origin_user_id = RemoteIdentity.where(user_id: auth_strategy.user,
                                                      auth_source: storage.oauth_client,
                                                      integration: storage)
                                               .pick(:origin_user_id)
                if origin_user_id.present?
                  ServiceResult.success(result: origin_user_id)
                else
                  failure(code: :error,
                          data: StorageErrorData.new(source: caller),
                          log_message:
                            "No origin user ID or user token found. Cannot execute query without user context.")
                end
              when :sso_user_token
                origin_user_id = RemoteIdentity.where(user_id: auth_strategy.user,
                                                      auth_source: user.authentication_provider,
                                                      integration: storage)
                                               .pick(:origin_user_id)
                if origin_user_id.present?
                  ServiceResult.success(result: origin_user_id)
                else
                  failure(code: :error,
                          data: StorageErrorData.new(source: caller),
                          log_message:
                            "No origin user ID or user token found. Cannot execute query without user context.")
                end

```
Other places should be fine with new constraint. But maybe I am missing something.

Reply:
Essentially I didn't think enough about edge cases right now, and I'd be happy to delay that.

But on a general level:
* `uniqueness: { scope: %i[auth_source integration] }` means that it's possible that `user.remote_identities.where(storage:).count > 1`
* This means there could be two different `origin_user_id` values for the same user at a storage (user.remote_identities.where(storage:).pluck(:origin_user_id).uniq.count > 1`)
* This means if we have code like `RemoteIdentity.where(user:, integration:).pick(:origin_user_id)` it could pick the wrong user id

However, you already properly addressed the one case I know of in your example above, by making sure we pick the correct `auth_source`. And other places I see in this PR actually work correctly by iterating over all identities, regardless of the auth source. So it's probably fine I guess?

Reply:
That's my understanding as well.